### PR TITLE
Entrypoints accept references as arguments

### PR DESCRIPTION
**Pull request by kpob**

### CODE REVIEW

File: examples/src/balance_checker.rs [hunk 1]
```diff
@@ -5,14 +5,14 @@ pub struct BalanceChecker {}
 
 #[odra::module]
 impl BalanceChecker {
-    pub fn check_balance(&self, token: Address, account: Address) -> U256 {
-        TokenRef::at(token).balance_of(account)
+    pub fn check_balance(&self, token: &Address, account: &Address) -> U256 {
+        TokenRef::at(*token).balance_of(account)
```
Comment (zie1ony) [inline examples/src/balance_checker.rs]: Change `at` to accept `T: Into<Address>`. Or check how to use `AsRef`.

File: examples/src/docs/native_token.rs [hunk 2]
```diff
@@ -9,8 +9,8 @@ impl PublicWallet {
     #[odra(payable)]
     pub fn deposit(&mut self) {}
 
-    pub fn withdraw(&mut self, amount: Balance) {
-        contract_env::transfer_tokens(contract_env::caller(), amount);
+    pub fn withdraw(&mut self, amount: &Balance) {
+        contract_env::transfer_tokens(contract_env::caller(), *amount);
```
Comment (zie1ony) [inline examples/src/docs/native_token.rs]: `transfer_tokens` could accepet `Into<Address>` and `Into<Balance>`. Or `AsRef`.

File: modules/src/erc1155_receiver.rs [hunk 3]
```diff
@@ -9,41 +9,38 @@ pub struct Erc1155Receiver {}
...
-        #[allow(unused_variables)] operator: Address,
-        #[allow(unused_variables)] from: Address,
-        #[allow(unused_variables)] token_id: U256,
-        #[allow(unused_variables)] amount: U256,
-        #[allow(unused_variables)] data: Option<Bytes>
+        #[allow(unused_variables)] operator: &Address,
+        #[allow(unused_variables)] from: &Address,
+        #[allow(unused_variables)] token_id: &U256,
+        #[allow(unused_variables)] amount: &U256,
+        #[allow(unused_variables)] data: &Option<Bytes>
```
Comment (zie1ony) [inline modules/src/erc1155_receiver.rs:16]: to `Option<Bytes>`.

File: modules/src/erc1155_receiver.rs [hunk 4]
```diff
@@ -9,41 +9,38 @@ pub struct Erc1155Receiver {}
 impl Erc1155Receiver {
     pub fn on_erc1155_received(
         &mut self,
-        #[allow(unused_variables)] operator: Address,
-        #[allow(unused_variables)] from: Address,
-        #[allow(unused_variables)] token_id: U256,
-        #[allow(unused_variables)] amount: U256,
-        #[allow(unused_variables)] data: Option<Bytes>
+        #[allow(unused_variables)] operator: &Address,
```
Comment (zie1ony) [inline modules/src/erc1155_receiver.rs:12]: Revert back to values in this file.

File: modules/src/erc721/extensions/erc721_receiver.rs [hunk 5]
```diff
@@ -4,9 +4,9 @@ use odra::types::{Address, U256};
 pub trait Erc721Receiver {
     fn on_erc721_received(
         &mut self,
-        operator: Address,
-        from: Address,
-        token_id: U256,
-        data: Option<Vec<u8>>
+        operator: &Address,
```
Comment (zie1ony) [inline modules/src/erc721/extensions/erc721_receiver.rs:7]: Dont use refs here.

File: modules/src/mock/pauseable.rs [hunk 6]
```diff
@@ -34,7 +34,8 @@ impl PauseableCounter {
 
 impl PauseableCounter {
     fn raw_increment(&mut self) {
-        self.value.set(self.value.get_or_default() + 1);
+        let new_value = self.value.get_or_default() + 1;
```
Comment (zie1ony) [inline modules/src/mock/pauseable.rs:37]: revert back

**Review by zie1ony — CHANGES_REQUESTED** [review]

File: modules/src/erc20.rs [hunk 7]
```diff
@@ -36,40 +36,40 @@ impl Erc20 {
         self.decimals.set(decimals);
 
         if let Some(initial_supply) = initial_supply {
-            self.total_supply.set(initial_supply);
-            self.balances.set(&caller, initial_supply);
+            self.total_supply.set(*initial_supply);
+            self.balances.set(&caller, *initial_supply);
 
-            if initial_supply > U256::zero() {
+            if *initial_supply > U256::zero() {
```
Comment (zie1ony) [inline modules/src/erc20.rs]: Use `is_zero`

File: modules/src/erc20.rs [hunk 8]
```diff
@@ -36,40 +36,40 @@ impl Erc20 {
         self.decimals.set(decimals);
 
         if let Some(initial_supply) = initial_supply {
```
Comment (zie1ony) [inline modules/src/erc20.rs]: Could you deref in `if let`?

File: odra-casper/backend/src/casper_env.rs [hunk 9]
```diff
@@ -265,13 +278,55 @@ where
         Err(_) => runtime::revert(ApiError::ValueNotFound)
     }
 }
+pub const DICTIONARY_ITEM_KEY_MAX_LENGTH: usize = 64;
 
 pub fn save_value<T: OdraType>(key: &str, value: T) {
     let state_uref = get_state_uref();
-    storage::dictionary_put(state_uref, key, value);
+
```
Comment (zie1ony) [inline odra-casper/backend/src/casper_env.rs]: Revert back.

File: odra-casper/codegen/src/arg.rs [hunk 10]
```diff
@@ -2,17 +2,26 @@ use odra_types::contract_def::Argument;
...
     fn to_tokens(&self, tokens: &mut TokenStream) {
         let args = self.0;
         args.iter().for_each(|arg| {
             let arg_ident = format_ident!("{}", arg.ident);
-
-            tokens.append_all(quote! {
-                let #arg_ident = odra::casper::casper_contract::contract_api::runtime::get_named_arg(stringify!(#arg_ident));
-            });
+            match &arg.ty {
+                odra_types::Type::Slice(ty) => {
```
Comment (zie1ony) [inline odra-casper/codegen/src/arg.rs:15]: What happens on args like `a: &[u8; 32]`?

File: odra-casper/codegen/src/ty/odra.rs [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+        assert_eq!(tokens.to_string(), "[u8 ; 32u32]");
+    }
+
+    #[test]
+    fn test_slice() {
+        let ty = Type::Slice(Box::new(Type::Bool));
+        let odra_type = OdraType(&ty);
+        let mut tokens = TokenStream::new();
+        odra_type.to_tokens(&mut tokens);
+        assert_eq!(tokens.to_string(), "Vec < bool >");
```
Comment (zie1ony) [inline odra-casper/codegen/src/ty/odra.rs]: Ask ChatGPT to use `quote`.

**Review by zie1ony — APPROVED** [review]

File: modules/src/erc1155_receiver.rs (hunk 3 above)
Comment (kpob) [reply modules/src/erc1155_receiver.rs:16]: Generally `&Option<T>` or `Option<&T>` should be used. Option is a zero-cost abstraction but is not a pointer, so its memsize is `size_of::<T>() + 1`

**Pull request merged**